@here/harp-mapview/lib/ClipPlanesEvaluator.ts

```diff
@@ -0,0 +1,179 @@
+
+    constructor(
+        readonly nearMin: number = 0.1,
+        readonly nearMultiplier: number = 0.1,
+        readonly nearFarMultiplier = 50.0,
+        readonly farOffset = 200.0
+    ) {
+        assert(nearMin > 0);
+        assert(nearFarMultiplier >= 0);
+        assert(farOffset >= 0);
+        this.farMin = nearMin * nearFarMultiplier + farOffset;
+    }
+
+    evaluateClipPlanes(camera: Camera, projection: Projection): { near: number; far: number } {
+        let nearPlane: number = this.nearMin;
+        let farPlane: number = this.farMin;
+        if (projection.type === ProjectionType.Spherical) {
+            // near and far plane for a set up where
+            // the camera is looking at the center of the scene.
+            const r = EarthConstants.EQUATORIAL_RADIUS;
+            const d = camera.position.length();
+            const alpha = Math.asin(r / d);
+            // Extract X, Y, Z axes into tmp vectors array.
+            camera.matrixWorld.extractBasis(
+                this.m_tmpVectors[0],
+                this.m_tmpVectors[1],
+                this.m_tmpVectors[2]
+            );
+            // Setup quaternion based on X axis.
+            this.m_tmpQuaternion.setFromAxisAngle(this.m_tmpVectors[0], alpha);
+            // Acquire forward vector based on Z axis reversed (keep it in tmpVectors[2]).
+            const fwd = this.m_tmpVectors[2].negate();
+            const p = camera.position.clone();
+            p.addScaledVector(
+                fwd.clone().applyQuaternion(this.m_tmpQuaternion),
+                Math.sqrt(d * d - r * r)
+            );
+            farPlane = p
+                .clone()
+                .sub(camera.position)
+                .dot(fwd);
+            const bias = 2000; // TODO: generalize.
+            nearPlane = Math.max(this.nearMin, projection.groundDistance(camera.position) - bias);
+        } else {
+            const groundDistance = projection.groundDistance(camera.position);
+            nearPlane = Math.max(this.nearMin, groundDistance * this.nearMultiplier);
+            // Will be already clamped to minFar due to clamping above.
+            farPlane = nearPlane * this.nearFarMultiplier + this.farOffset;
+        }
+        return { near: nearPlane, far: farPlane };
+    }
+}
+
+/**
+ * Evaluates camera clipping planes based on the eye vector angle with surface of the ground.
+ */
+export class TiltBasedClipPlanesEvaluator extends GroundBasedClipPlanesEvaluator {
+    private static readonly EYE_INVERSE = new THREE.Vector3(0, 0, -1);
+
+    evaluateClipPlanes(camera: Camera, projection: Projection): { near: number; far: number } {
+        const clipPlanes = super.evaluateClipPlanes(camera, projection);
+
+        // Tilt angle will be taken into consideration only for non-spherical projections,
+        // otherwise take super class values.
+        if (projection.type !== ProjectionType.Spherical) {
+            camera.getWorldDirection(this.m_tmpVectors[0]);
```

Comment:
```m_tmpVectors``` array does not exist in this class

Reply:
This is super class property, according to language specs:
https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#4.9.2
in this case there is no difference between super. or this. accessor, because there are no overrides in the derived class instance.

Reply:
But I see other bug here. m_tmpVectors are not initialized - array with nulls. So maybe I misunderstand you or you helped me to find bug anyway :).